File: Data Structures & Algorithms/binary-tree-diameter/submission-10.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_val = 0

        def find_diam(node):
            nonlocal max_val
            if not node:
                return 0
            
            left_val =  find_diam(node.left)
            right_val = find_diam(node.right)
            max_val = max(max_val, left_val + right_val)
            print(node.val, left_val, right_val, max_val)
            return 1 + max(left_val, right_val)
            
        find_diam(root)
        return max_val

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-10.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(root, p, q) {
        let isFound = false; 
        let response; 

        const findLCA = (root, p, q) => {
            if(!root) return;
            if(isFound) return;
            if(root.left && root.right && root.left.val == p.val && root.right.val == q.val || root.left.val == q.val && root.right.val == p.val){
                response = root; 
                isFound = true; 
                return;
            }

            if(root.val == p.val || root.val == q.val){
                response = root; 
                isFound = true; 
                return;
            }

            findLCA(root.left, p, q)
            findLCA(root.right, p, q)
        }

        findLCA(root, p, q)
        return response
    }
}
